### ISSUE TITLE
2 Stoploss questions 

### ISSUE THREAD

**Issue by FahadAziz** [freqtrade/freqtrade#8776]
## Describe your environment

  * Operating system: Ubuntu 22.10 x64
  * Python Version: Python 3.10.7
  * CCXT version: ccxt==3.1.5
  * Freqtrade Version: freqtrade 2023.5.1
  * 
## Your question

Hi, I have few questions regarding existing trades as I am trying to mimic how the backtest work in a live trading and I faced two issues with two of my approaches. 

Since Backtest would not look at intra candles I figured I would stop the bot from exiting trading within the same candle with this code:

```
    def confirm_trade_exit(self, pair: str, trade: Trade, order_type: str, amount: float,
                           rate: float, time_in_force: str, exit_reason: str,
                           current_time: datetime, **kwargs) -> bool:

   
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=pair,
                                                                timeframe=self.timeframe)
        new_exitprice = dataframe['date'].iat[-1]


        if exit_reason == 'trailing_stop_loss' and ((current_time - trade.open_date_utc) < timedelta(minutes=15)):
            return False
        
        if exit_reason == 'roi' and ((current_time - trade.open_date_utc) < timedelta(minutes=15)):
            return False

        

        return True
```

This code works fine until the bot registers ROI or Trailing Stoploss exit and sticks to it. If the trade hit my trailing stop intra candle it will try to exit but the code will prevent it from exiting until it completes a full candle. the problem with this approach is, lets say that the bot hit the trailing target and then dropped to hit the stop loss, the bot will not see that which will result in a huge loss as the bot will continuously try to exit as a trailing exit. I do not know if this is intended or a bug but it is worth looking at. 

To overcome this issue I used stop loss on exchange which works perfectly until I noticed that it will create emergency exits for a trade that is hitting the trailing or ROI target. I know this is how the code should works. I tried making "stoploss_on_exchange_interval" longer but it would not be ideal as it will interfere with my code above and would not act as intended. 

I exhausted all resources looking for answers to overcome this issue. Ideas and direction would be appreciated.

Thanks

**Comment by stash86** [comment]
While backtest can't replicate intra-candle movement, if you are using the "simple" trailing setting, during backtest, FT will check whether the target profit would fall between open-high area, and gonna be triggered if it would.

same thing for ROI.

so there is no reason to do such tricks like what you did. The only time you need to do the trick above is if you are using custom_exit.

**Comment by stash86** [comment]
Forget to answer your other question.

Once a trailing_stop_loss triggered, there is no such thing as "stoploss" anymore, since you have move the stoploss from its initial position. Not sure if it can be considered as a bug or not, but it's working as intended.

**Comment by xmatthias** [comment]
> Since Backtest would not look at intra candles

that's not entirely true.
your code implies that you're (probably) using 15m candles. 
You could backtest with `--timeframe-detail 1m` - which would give you 1m level detail (assumes data is available/downloaded, obviously).
While not 100% intra-candle - 1m candles will almost always give you a pretty good picture if you're actually running on a higher timeframe.


For your logic - it seems to simply prevent trailing stop / roi exits within the first 15 minutes.
After that, the logic seems to be inactive - but it also seems counter-intuitive to prevent exits initially. If your ROI is hit - why ignore it for the first 15 minutes, and not simply set the initial ROI target higher?

**Comment by FahadAziz** [comment]
@stash86 Thank you for your input. My strategy is a volatility based strategy and will result in live trading. bot will always get out early that's why I want at least a full candle to be complete before bot try to close the trade without risking initial stop loss to be missed.

@xmatthias I am aware of this but it would not fully mimic live trading. I agree it will give a better picture but I am more worried about mimicking that into live trading, I want a full candle to complete before exiting any trade while adhering to stop loss. reason I want this logic is basically I trade with leverage and bot is exiting trades early intra-candle through trailing stop loss which will result in backtests completely misleading. 

This code seems to do the trick so far with stop loss on exchange active. I am yet to see how it would behave when it hits ROI 

``` python
    use_custom_stoploss = True

    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:

        if current_profit > 0.03 and ((current_time - trade.open_date_utc) > timedelta(minutes=15)):
            return -0.01

        return -1
```

**Comment by xmatthias** [comment]
Effectively, what it really means is that your trailing stop parameters are too tight.
you're running with 1% trailing stop - assuming a leverage of 5x, that'll result in an effective trailing stoploss of 0.2% (1 / 5x)- which will almost certainly trigger immediately.

With such a tight stop, you'll also encounter problems actually placing the stoploss - which will result in more early exits, too.

---

when using custom_stoploss, i'd recommend to disable the regular trailing stop.
while it's technically allowed, it's highly discouraged - and the 2 will compete with each other

**Comment by FahadAziz** [comment]
Yes I would certainly mean that, but the problem I am facing is bot will exit early at 3% let's say but if it waited for the full candle to complete it would mean >%10. This is just an example which the backtest is capturing since exit will happen at the next candle and will result in a more favorable backtest results while in live trading it would miss on that. I disabled trailing stop as well.

So far placing the stop loss on exchange is not an issue it worked perfectly I am yet to give it some time to see how it would behave when a lot of trades on the book and when it hits ROI because I do not want any trade to exit intra.

**Comment by stash86** [comment]
Not sure if I get the logic. If any exit logic triggered within few minutes, why have to wait for full candle? There is no guarantee the price will be in your favor after the trade candle closed. You are putting yourself on a bigger risk than necessary. If my roi reached after 10 seconds, so be it. Because there is no guarantee that the price will stay up there

**Comment by FahadAziz** [comment]
The whole exercise is to replicate to a certain extent what I see in backtest result. Exiting intra candle would mean doing backtests is useless if what I see in the backtest is not replicated in terms of logic into live trading. maybe this is not applicable to spot trading but I am trading futures with leverage and I am seeing that most of trades are exited early so I am missing on a lot of gains "seen on backtest but not in live trades". 

Update and a question: I know this was obvious to you gents without trying it live. But the code works perfectly now with stop loss on exchange enabled for "stop loss and trailing stop loss". is it safe to assume that I can play with `custom_exit` function to find a way I can stop ROI exit intra candle? how would I go about it while reading directly from ROI table?

**Comment by stash86** [comment]
And as said above, it's not an issue for roi and trailing stoploss. If you have discrepancy between backtest and live, usually it because you have different entry rate. Backtest always assume entry rate equal open rate. In live, esp since you are trading on volatile pairs, most likely you gonna get higher rate than open rate, which means your tricks done nothing to prevent the difference.

second reason is slippage. Roi or trailing triggered, but then the price dropping hard. You get exit rate that is lower than what you supposed to get. There is no slippage in backtest. Your trick also won't prevent this.

so in the end, what you are doing is pointless in term of eliminating (what you think is) problems.

**Comment by stash86** [comment]
> Yes I would certainly mean that, but the problem I am facing is bot will exit early at 3% let's say but if it waited for the full candle to complete it would mean >%10. This is just an example which the backtest is capturing since exit will happen at the next candle and will result in a more favorable backtest results while in live trading it would miss on that. I disabled trailing stop as well.
> 
> So far placing the stop loss on exchange is not an issue it worked perfectly I am yet to give it some time to see how it would behave when a lot of trades on the book and when it hits ROI because I do not want any trade to exit intra.

This is what we call hindsight 20/20. While that specific trade maybe went rocket, there would also be trades that went 3-4% intra-candle, before going south before the candle closed. Then you are missing out on potential profit and ends up with losses.

Please read about "risk management". It's more important than "profit maximizing"

**Comment by xmatthias** [comment]
I think your approach is wrong - You're trying to force live to be have like backtest - while in reality, it won't (backtesting takes some [assumptions](https://www.freqtrade.io/en/stable/backtesting/#assumptions-made-by-backtesting) - which will not hold in live, also not with "limitations" to avoid intra-candle action.

The logic of "i don't want intra-candle orders" is like a child trying to hide by closing it's eyes and screaming "nobody can see me".
it may work as long as all others play along - but will fail hard should it REALLY need to hide for whatever reason.

I therefore think you're moving in the wrong direction - though as it's not my money, i don't see it as my task to prevent you from losing it (you've been warned, and we have warnings in the docs, which will suffice).

> is it safe to assume that I can play with custom_exit function to find a way I can stop ROI exit intra candle? how would I go about it while reading directly from ROI table?

It's up to you to come up with logics that work for your strategy.
you can access the ROI table with `self.minimal_roi`. 
It's however up to you to replicate the logic if you must - it's neither something we recommend, nor something we'll support.
